logstash-core/lib/logstash/agent.rb

```diff
@@ -26,26 +39,112 @@ def execute
     sigint_id = trap_sigint()
     sigterm_id = trap_sigterm()
 
-    @pipelines.each {|_, p| p.run } # blocking operation. works now because size <= 1
+    start_background_services
+
+    @pipelines.each { |_, p| Thread.new { p.run } }
+    start_webserver
+
+    sleep(1) while true
     return 0
   rescue => e
     @logger.fatal I18n.t("oops", :error => e)
     @logger.fatal e.backtrace if @logger.debug? || $DEBUGLIST.include?("stacktrace")
     return 1
   ensure
+    stop_background_services
+
     Stud::untrap("INT", sigint_id) unless sigint_id.nil?
     Stud::untrap("TERM", sigterm_id) unless sigterm_id.nil?
   end # def execute
 
   def add_pipeline(pipeline_id, config_str, settings = {})
-    @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings.merge(:pipeline_id => pipeline_id))
+    settings.merge!(:pipeline_id => pipeline_id,
+                    :metric => metric)
+
+    @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings)
   end
 
   def node_uuid
     @node_uuid ||= SecureRandom.uuid
   end
 
+  # Calculate the Logstash uptime in milliseconds
+  #
+  # @return [Fixnum] Uptime in milliseconds
+  def uptime
+    ((Time.now.to_f - started_at.to_f) * 1000.0).to_i
+  end
+
   private
+
+  def start_webserver
+    options = { :debug => debug }
+    @webserver = LogStash::WebServer.new(@logger, options)
+    Thread.new(@webserver) do |webserver|
+      webserver.run
+    end
+  end
+
+  def stop_webserver
+    @webserver.stop
+  end
+
+  def start_background_services
+    if collect_metric?
+      @logger.debug("Agent: Starting metric periodic pollers")
+      @periodic_pollers.start
+    end
+  end
+
+  def stop_background_services
+    if collect_metric?
+      @logger.debug("Agent: Stopping metric periodic pollers")
+      @periodic_pollers.stop
+    end
+  end
+
+  def configure_metric
+    if collect_metric?
+      @logger.debug("Agent: Configuring metric collection")
+      @metric = LogStash::Instrument::Metric.create
+      add_metric_pipeline
+    else
+      @metric = LogStash::Instrument::NullMetric.new
+    end
+
+    @periodic_pollers = LogStash::Instrument::PeriodicPollers.new(metric)
+  end
+
+  def collect_metric?
+    @collect_metric
+  end
+
+  # Add a new pipeline sitting next to the main pipeline,
+  # This pipeline should only contains one input: the `metrics`
+  # and multiple shippers.
+  def add_metric_pipeline
+    @logger.debug("Agent: Adding metric pipeline")
+
+    metric_pipeline_config =<<-EOS
+      input {
+        metrics {}
+      }
+      output {
+        elasticsearch {
+          flush_size => 1
+          hosts => "127.0.0.1"
+          index => "metrics-%{YYYY.MM.dd}"
```

Comment:
`index => "metrics-%{YYYY.MM.dd}"` should be     `index => "metrics-%{+YYYY.MM.dd}"`